### PR TITLE
DM relay - minor fixes

### PR DESCRIPTION
**Pull request by lemonsaurus**
This hotfix PR makes three minor changes - We restore user caching, add the user ID to the author name, and add some stats.

## User caching
We previously had a feature to allow a `!reply` without specifying the target user. This was removed because we were afraid that the number of DMs would be so great that the risk of replying to the wrong user was high.

As it turns out, this traffic is much lower than anticipated and the risk of replying to the wrong user appears to be low. That, coupled with the fact that we shouldn't really be using `!reply` for anything serious, has made me decide to restore this functionality.

## Adding the user ID to the username
When we get a DM, there's no actual user mention since it's relayed via a webhook. That means that a moderator would have to search for that user and figure out what their ID is in order to do stuff like infraction searches.

To make this easier, I'm adding the user ID to the author name:
image

## Stats!
We're now incrementing `dm_relay.dm_sent` whenever we send a DM, and `dm_relay.dm_received` whenever we receive one. This should allow us to track how many DMs the bot sends and receives.

### CODE REVIEW

File: bot/cogs/dm_relay.py [hunk 1]
```diff
@@ -17,30 +19,45 @@
...
+                self.bot.stats.incr("dm_relay.dm_sent")
+                return
+            elif last_dm_user:
+                await last_dm_user.send(message)
+                await ctx.message.add_reaction("✅")
+                self.bot.stats.incr("dm_relay.dm_sent")
+                return
+            else:
+                log.debug("This bot has never gotten a DM, or the RedisCache has been cleared.")
+                await ctx.message.add_reaction("❌")
```
Comment (SebastiaanZ) [inline bot/cogs/dm_relay.py:60]: Why don't we do something like:

```suggestion
        if not member:
            user_id = await self.dm_cache.get("last_user")
            member = ctx.guild.get_member(user_id) if user_id else None

        # If we still don't have a Member at this point, give up
        if not member:
            log.debug("This bot has never gotten a DM, or the RedisCache has been cleared.")
            await ctx.message.add_reaction("❌")
            return

        try:
            await member.send(message)
        except discord.errors.Forbidden:
            log.debug("User has disabled DMs.")
            await ctx.message.add_reaction("❌")
        else:
            await ctx.message.add_reaction("✅")
            self.bot.stats.incr("dm_relay.dm_sent")
```

File: bot/cogs/dm_relay.py [hunk 2]
```diff
@@ -17,30 +19,45 @@
...
+
     def __init__(self, bot: Bot):
         self.bot = bot
         self.webhook_id = constants.Webhooks.dm_log
         self.webhook = None
         self.bot.loop.create_task(self.fetch_webhook())
 
     @commands.command(aliases=("reply",))
-    async def send_dm(self, ctx: commands.Context, member: discord.Member, *, message: str) -> None:
+    async def send_dm(self, ctx: commands.Context, member: Optional[discord.Member], *, message: str) -> None:
```
Comment (MarkKoz) [inline bot/cogs/dm_relay.py:32]: This converter will convert by name and nickname too. So if someone happens to have a common word as a name, you'll end up DMing them.

Comment (lemonsaurus) [reply bot/cogs/dm_relay.py:32]: That's a very good point. What I really want is a converter that only does mentions and userIDs. Maybe I'll need to make a custom one?

File: bot/cogs/dm_relay.py (hunk 1 above)
Comment (lemonsaurus) [reply bot/cogs/dm_relay.py]: Sure.

File: bot/cogs/dm_relay.py (hunk 2 above)
Comment (lemonsaurus) [reply bot/cogs/dm_relay.py]: Fixed in 403572b

**Pull request merged**